### ISSUE TITLE
`convox deploy` suddenly taking a really long time

### ISSUE THREAD

**Issue by ndbroadbent** [convox/rack#2531]
`convox deploy` just started taking a really long time, but my instances aren't using much CPU (< 10%). I'm running 3x `c4.large` instances.

There was an `ARG` step in my Docker build that took over 10 minutes, and I'm not sure what is making my builds run so slowly.

Even after rebooting all 3 instances, this step is taking over 2 minutes:

```
Step 15/47 : ARG SECRET_KEY_BASE
 ---> Running in 00e1999a5ba7

... wait 2 minutes ...

 ---> 53c0d9f219a5
Removing intermediate container 00e1999a5ba7
```

```
$ convox ps info 7f7d79a7b6fe
Id       7f7d79a7b6fe
Name     build
Release  20180307152136
CPU      0.00%
Memory   0.00%
Started  1 minute ago
```

It just seems like it's sitting there and doing nothing. Could this be something to do with EFS?

Note: This was still happening even after I stopped all of my web and worker containers.

P.S. I tried to run `exec` to get into the build container, so that I could use `ps` to see how much memory/CPU was being used, but that caused an error:

```
$ convox exec 7f7d79a7b6fe bash
oci runtime error: exec failed: container_linux.go:265: starting container process caused "exec: \"/convox-env\": stat /convox-env: no such file or directory"
```

**Comment by ddollar** [comment]
Did you figure out the cause here?

**Comment by ndbroadbent** [comment]
Sorry I didn't figure it out, but I upgraded to `c5.large` instances and it seems to be faster now. 

Also I realized this is more of a support ticket and I'm not paying for a Pro subscription, so I felt bad about opening too many issues. But I think I'm all set now, everything seems to be stable and I'm moving over my production servers. Thanks for all your help!

**Comment by mwarkentin** [comment]
@ndbroadbent might want to try using a dedicated build instance as well

**Comment by ndbroadbent** [comment]
@mwarkentin Thanks, I didn't see [the docs about dedicated build instances](https://convox.com/docs/advanced-build-options/#dedicated-build-instance). (Posting the link in case someone else finds this.)

I just saw the new release with support for [running timers on Fargate](https://github.com/convox/rack/pull/2527). Would it be possible to run builds on Fargate?

**Comment by mwarkentin** [comment]
I don’t think fargate has any support for building docker containers, just running them.

**Comment by ndbroadbent** [comment]
Ah, I see.

I've started a new build instance (`c5.large`), but Docker is still taking ~15 seconds to run each `ARG` step:

![docker-slow](https://user-images.githubusercontent.com/139536/37213408-8626a00c-23e4-11e8-9522-67221a244897.gif)

I have 20 build args, so this is taking about 5 minutes. The full `convox deploy` is taking over 16 minutes:

```
$ time convox deploy
...
real	16m11.575s
```

Building the same image on my MacBook Pro is a lot faster:

![faster-local-build](https://user-images.githubusercontent.com/139536/37214218-ed5e278e-23e6-11e8-953b-911160b2abd6.gif)

```
$ time docker build --no-cache ...
...
real	2m21.998s
```

I noticed these options in `convox rack params`:

```
BuildCpu                 0
BuildImage
BuildInstance            c5.large
BuildMemory              1000
BuildVolumeSize          200
```

I tried `convox rack params set BuildMemory=2048 BuildCpu=2048`, but that didn't help. 

I couldn't find any docs for the `BuildVolumeSize` parameter. My base image is 1.27GB, so maybe that's the problem?

**Comment by ndbroadbent** [comment]
Hi again, I've been reading a lot about how the `devicemapper` Storage Driver can cause Docker to be very slow. That's what my instances are using: 

```
[ec2-user@ip-10-0-3-122 ~]$ docker info | grep "Storage Driver"
Storage Driver: devicemapper
```

* https://stackoverflow.com/a/35496394/304706
* https://www.danstutzman.com/2014/07/speed-up-docker-on-ec2.html

Is there any way to switch to `aufs`?

**Comment by m1ch3lp3r3z** [comment]
I was having the same issue, even more severe, my deployments on our development rack could take up to 20 minutes, :0. I had t2.small instances, but a couple of months ago that was just fine (about same time I did some version updates). I just changed the type of my EC2 instances to t2.medium, not really big difference but it just rebuilt all (3) instances and it's really fast (~ 2 min) now. I did not check the instances directly, but wondering if it could be a space problem, instances running out of space. How convox manages any downloaded images, are images not being used deleted?. Thanks.
